File: frontend/client-api/orchestrator.ts
import { fetchJson } from "@/client-api/client";
import type { ActionQueueItem, OrchestratorAnalytics, OrchestratorStatus } from "@/types/app";

export function getActionQueue(params: { status?: string; limit?: number } = {}): Promise<{ queue: ActionQueueItem[] }> {
  const search = new URLSearchParams();
  if (params.status) search.set("status", params.status);
  if (params.limit) search.set("limit", String(params.limit));
  const query = search.toString();
  return fetchJson(`/api/orchestrator/queue${query ? `?${query}` : ""}`);
}

export function getProjectOrchestratorStatus(projectId: number): Promise<{ orchestrator: OrchestratorStatus }> {
  return fetchJson(`/api/orchestrator/projects/${projectId}/status`);
}

export function startProjectOrchestrator(projectId: number): Promise<{ orchestrator: { created: number; project_id: number; queue: ActionQueueItem[] } }> {
  return fetchJson(`/api/orchestrator/projects/${projectId}/start`, { method: "POST" });
}

export function runProjectOrchestrator(projectId: number): Promise<{ orchestrator: OrchestratorStatus & { executed: number } }> {
  return fetchJson(`/api/orchestrator/projects/${projectId}/run`, { method: "POST" });
}

export function retryQueueAction(actionId: number): Promise<{ action: ActionQueueItem }> {
  return fetchJson(`/api/orchestrator/actions/${actionId}/retry`, { method: "POST" });
}

export function getOrchestratorAnalytics(): Promise<{ orchestrator: OrchestratorAnalytics }> {
  return fetchJson("/api/orchestrator/analytics");
}

File: backend/app/routers/orchestrator.py
from fastapi import APIRouter, Depends, HTTPException, Query

from app.auth import ensure_not_maintenance_mode, require_roles
from app.db import get_db
from app.orchestrator import (
    build_orchestrator_analytics,
    enqueue_project_orchestration,
    get_project_orchestrator_status,
    list_action_queue,
    retry_queue_action,
    run_project_orchestrator,
)
from app.rate_limit import rate_limit_dependency

router = APIRouter(prefix="/api/orchestrator", tags=["orchestrator"])


@router.get("/queue")
async def get_queue(
    status: str = Query("", max_length=40),
    limit: int = Query(100, ge=1, le=200),
    user: dict = Depends(require_roles("admin", "manager")),
) -> dict:
    with get_db() as db:
        return {"queue": list_action_queue(db, status=status, limit=limit, user_id=int(user["id"]))}


@router.get("/analytics")
async def get_analytics(user: dict = Depends(require_roles("admin", "manager"))) -> dict:
    with get_db() as db:
        return {"orchestrator": build_orchestrator_analytics(db, int(user["id"]))}


@router.get("/projects/{project_id}/status")
async def get_project_status(project_id: int, user: dict = Depends(require_roles("admin", "manager", "member", "viewer"))) -> dict:
    with get_db() as db:
        return {"orchestrator": get_project_orchestrator_status(db, project_id, int(user["id"]))}


@router.post("/projects/{project_id}/start")
async def start_project_orchestrator(
    project_id: int,
    user: dict = Depends(require_roles("admin", "manager", "member")),
    _: None = Depends(rate_limit_dependency("generation")),
) -> dict:
    ensure_not_maintenance_mode()
    with get_db() as db:
        result = enqueue_project_orchestration(db, project_id=project_id, user_id=int(user["id"]))
        if not result.get("queue") and not result.get("created"):
            raise HTTPException(status_code=404, detail="Project not found.")
        return {"orchestrator": result}


@router.post("/projects/{project_id}/run")
async def run_project(
    project_id: int,
    user: dict = Depends(require_roles("admin", "manager", "member")),
    _: None = Depends(rate_limit_dependency("generation")),
) -> dict:
    ensure_not_maintenance_mode()
    with get_db() as db:
        enqueue_project_orchestration(db, project_id=project_id, user_id=int(user["id"]))
        result = run_project_orchestrator(db, project_id=project_id, user_id=int(user["id"]))
        if not result.get("queue"):
            raise HTTPException(status_code=404, detail="Project not found.")
        return {"orchestrator": result}


@router.post("/actions/{action_id}/retry")
async def retry_action(
    action_id: int,
    user: dict = Depends(require_roles("admin", "manager")),
    _: None = Depends(rate_limit_dependency("admin")),
) -> dict:
    ensure_not_maintenance_mode()
    with get_db() as db:
        action = retry_queue_action(db, action_id=action_id, user_id=int(user["id"]))
    if not action:
        raise HTTPException(status_code=404, detail="Action not found.")
    return {"action": action}
